code/obj/flock/structure/sapper.dm

```diff
@@ -0,0 +1,150 @@
+#define CHARGING_BITS "Speed up Flockbits"
+#define CHARGING_DRONES "Charge incapacitors"
+#define CHARGING_STRUCTURES "Enhance structures"
+
+/obj/flock_structure/sapper
+	icon_state = "sapper-off" // current sprites are placeholders
+	name = "sparking machine"
+	desc = "A strange structure supporting four coils. You can feel the electricity coming off of it"
+	flock_desc = "A multipurpose support structure that saps power from a local APC. It can be set to enchance Flockbits, Flockdrones, or structures."
+	flock_id = "Sapper"
+	health = 30
+	health_max = 30
+	repair_per_resource = 1.5
+	resourcecost = 100
+
+	passthrough = TRUE
+
+	var/mode = CHARGING_BITS
+	var/mode_cooldowns = list(CHARGING_BITS = 20 SECONDS, CHARGING_DRONES = 30 SECONDS, CHARGING_STRUCTURES = 30 SECONDS)
+
+	var/obj/machinery/power/apc/linked_apc = null
+
+	New(atom/location, datum/flock/F = null)
+		..()
+		src.linked_apc = src.try_link_apc()
+		if (!src.linked_apc)
+			src.icon_state = "sapper-off"
+
+		src.info_tag.set_info_tag("Mode: [src.mode]")
+
+		ON_COOLDOWN(src, CHARGING_BITS, src.mode_cooldowns[CHARGING_BITS])
+		ON_COOLDOWN(src, CHARGING_DRONES, src.mode_cooldowns[CHARGING_DRONES])
+		ON_COOLDOWN(src, CHARGING_STRUCTURES, src.mode_cooldowns[CHARGING_STRUCTURES])
+
+	building_specific_info()
+		return {"<span class='bold'>Mode:</span> [src.mode].
+				<br><span class='bold'>Linked power supply charge:</span> [src.linked_apc?.cell ? "[round(src.linked_apc.cell.charge / src.linked_apc.cell.maxcharge * 100)]%": "Not linked"]."}
+
+	process(mult)
+		if (!src.linked_apc)
```

Comment:
```suggestion
		if (QDELETED(src.linked_apc))
```
catches when the apc gets destryoed but we still hold a ref

Reply:
Question: What in your opinion should happen when the sapper's tile changes its area or when the sapper moves? I know neither of those seem particularly common, but they can happen (for example turf getting blown up, sapper pulled by singulo etc.). Currently the sapper stays linked to its previous APC, possible other behaviour would be to connect to the APC of its new area.

Reply:
I think if it moves and stays in the same area the APC can stay the same but yeah it could change APCs if moving to a new area, will change